src/pam_neototp.8
.TH "PAM_NEOTOTP" "8" "04/02/2015" "" ""
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.nh
.ad l
.SH "NAME"
pam_neototp \- Module for TOTP authentication using a YubiKey NEO(-N)
.SH "SYNOPSIS"
.HP \w'\fBpam_neototp\&.so\fR\ 'u
\fBpam_neototp\&.so\fR [\&.\&.\&.]
.SH "DESCRIPTION"
.PP
The \fBpam_neototp\fR \s-1PAM\s0 module allows for TOTP system authentication using a YubiKey NEO(-N). The module will try every possible way to access the YubiKey if this becomes necessary to assert successful two factor authentication for system administrators even if the system is quite broken. For a total of two users a working USB subsystem and a functioning YubiKey neo is sufficient, in all other cases some daemon(s) are required.
.PP
Please keep in mind that a YubiKey is not the fastest device and that the OATH applet on the YubiKey has a memory limitation of 2048 bytes (each entry consists of the UTF8 entry name and 9 additional bytes) for TOTP tokens. So try not to use a YubiKey on a multi user and/or busy server. It is ok, however, to use a YubiKey even on a busy server if YubiKey based TOTP is restricted to e.g. only the few administrative users.
.SH "OPTIONS"
.PP
\fBtry_first_pass\fR
.RS 4
Use password already entered as TOTP token, if specified, otherwise inquire TOTP token.
.RE
.PP
\fBuse_first_pass\fR
.RS 4
Require an already entered password which is used as TOTP token.
.RE
.PP
\fBalwaysok\fR
.RS 4
Never fail authentication - this is for testing only!
.RE
.PP
\fBconfig=\fR\fB\fIpath\fR\fR
.RS 4
The required configuration file. See the installed example (/etc/neototp.conf.example) for further information.
.RE
.PP
\fBdigits=\fR\fB\fInumber\fR\fR
.RS 4
The required number of digits for the TOTP token. Valid numbers are 6, 7 and 8 with 6 being the default. If the amount of characters entered exceeds this number the leading characters are assumed to be a prefix as in <prefix><token-digits>. Such a prefix is then used to create an OATH applet entry name in the form of <user>:<prefix>. This allows for multiple real users to share a single logon account but use different tokens. If no such prefix is given the entry name of the OATH applet must be identical to the logon name.
.RE
.PP
\fBwindow=\fR\fB\fInumber\fR\fR
.RS 4
The search window (number*30s) from the current time to find a matching TOTP token. The valid range is 0 to 5 with 0 being the default. One should actually use a search window of at least 1 to allow for some clock drift on the client's side.  This is especially true if the client side user has to enter the token number manually which will take some time.
.RE
.PP
\fBcachedb=\fR\fB\fIpath\fR\fR
.RS 4
The optional pathname of the GDBM database used for successful authentication cacheing. This database should be deleted during system boot so it is best located in a tmpfs filesystem. If this option is not specified no authentication cacheing takes place.
.RE
.PP
\fBcachehosts=\fR\fB\fIpath\fR\fR
.RS 4
The pathname of an optional configuration file specifying the remote hosts for which successful authentiction cacheing is enabled. See the installed example (/etc/authcache.conf.example) for details.
.RE
.PP
\fBretrigger\fR
.RS 4
In case of renewed server access update the cached last access time. If this option is not specified and cacheing of valid accesses is enabled the cache entry expires exactly after the cache validity time specified. Note that this option is ignored when cacheing is not activated.
.RE
.PP
\fBvalid=\fR\fB\fInumber\fR\fR
.RS 4
The amount of seconds a successful authentication is cached. During the cached period no further TOTP authentication is requested. The valid range is 1 to 86400 seconds with 1800 seconds being the default. Note that this option is ignored when cacheing is not activated.
.RE
.PP
\fBreplaydb=\fR\fB\fIpath\fR\fR
.RS 4
The optional pathname of the GDBM database used for authentication replay detection. This database should be deleted during system boot so it is best located in a tmpfs filesystem. If this option is not specified no authentication replay detection takes place.
.RE
.PP
\fBreplayok=\fR\fB\fIpath\fR\fR
.RS 4
The pathname of an optional configuration file specifying the remote hosts for which authentiction replay detection is disabled. See the installed example (/etc/replayhosts.conf.example) for details.
.RE
.PP
\fBsocket=\fR\fB\fIpath\fR\fR
.RS 4
The optional pathname to the totpd daemon unix domain socket. Tf not specified or if the daemon is not running this module will try to access the YubiKey directly first via CCID and then in case of the user being associated to one of the two classic slots directly via USB.
.SH "MODULE TYPES PROVIDED"
.PP
All module types (\fBaccount\fR, \fBauth\fR, \fBpassword\fR and \fBsession\fR) are provided, though \fBaccount\fR and \fBsession\fR currently do nothing whereas \fBpassword\fR will always deny (token change is not possible this way).
.SH "RETURN VALUES"
.RE
.PP
PAM_SERVICE_ERR
.RS 4
Configuration error.
.RE
.PP
PAM_SUCCESS
.RS 4
Successful authentification.
.RE
.PP
PAM_AUTH_ERR
.RS 4
Authentication failure.
.RE
.PP
PAM_AUTHINFO_UNAVAIL
.RS 4
Authentication information not available, e.g. unknown user.
.RE
.PP
PAM_IGNORE
.RS 4
Always returned by \fBaccount\fR and \fBsession\fR.
.RE
.PP
PAM_PERM_DENIED
.RS 4
Always returned by \fBpassword\fR.
.RE
.SH "SEE ALSO"
.PP
\fBtotpd\fR(1),
\fBpam.d\fR(5),
\fBpam\fR(8)
.SH "AUTHOR"
.PP
pam_neototp was written by Andreas Steinmetz.


src/pam_neototp.c
/*
 * This file is part of the neosctotp project, which allows a YubiKey NEO(-N)
 * to be used on a client or on a server for TOTP two factor authentication.
 *
 * Copyright (c) 2015 Andreas Steinmetz, ast@domdv.de
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <sys/socket.h>
#include <sys/un.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <gdbm.h>
#include <arpa/inet.h>
#include <libneosc.h>

#include "config.h"
#define COMM_UNIX
#include "client.h"
#include "pamcommon.h"
#include "dbstuff.h"
#include "neoauth.h"

PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh,int flags,int argc,
	const char **argv)
{
	int i;
	int r;
	int try_first_pass=0;
	int use_first_pass=0;
	int alwaysok=0;
	int mode=0;
	int digits=6;
	int window=0;
	int valid=1800;
	time_t t;
	const char *user=NULL;
	const char *pass=NULL;
	const char *config=config_default;
	const char *replaydb=NULL;
	const char *cachedb=NULL;
	const char *cmp=NULL;
	const char *rcmp=NULL;
	const char *sock=NULL;
	char name[256];

	for(i=0;i<argc;i++)if(!strcmp(argv[i],"try_first_pass"))
		try_first_pass=1;
	else if(!strcmp(argv[i],"use_first_pass"))use_first_pass=1;
	else if(!strcmp(argv[i],"alwaysok"))alwaysok=1;
	else if(!strcmp(argv[i],"retrigger"))mode=1;
	else if(!strncmp(argv[i],"digits=",7))digits=atoi(&argv[i][7]);
	else if(!strncmp(argv[i],"window=",7))window=atoi(&argv[i][7]);
	else if(!strncmp(argv[i],"valid=",6))valid=atoi(&argv[i][6]);
	else if(!strncmp(argv[i],"config=",7))config=&argv[i][7];
	else if(!strncmp(argv[i],"replaydb=",9))replaydb=&argv[i][9];
	else if(!strncmp(argv[i],"cachedb=",8))cachedb=&argv[i][8];
	else if(!strncmp(argv[i],"cachehosts=",11))cmp=&argv[i][11];
	else if(!strncmp(argv[i],"replayok=",9))rcmp=&argv[i][9];
	else if(!strncmp(argv[i],"socket=",7))sock=&argv[i][7];

	if(config_parse((char *)config,0)||digits<6||digits>8||serial<0||
		window<0||window>5||valid<=0||valid>86400)
	{
		r=PAM_SERVICE_ERR;
		goto out;
	}

	if(cachedb&&(r=cachecheck(pamh,cachedb,cmp,valid,mode))==PAM_SUCCESS)
		goto out;

	if((r=preprocess(pamh,&user,&pass,name,sizeof(name),try_first_pass,
		use_first_pass,digits))!=PAM_SUCCESS)goto out;

	t=time(NULL);
	if(sock&&!unxclient(sock,t,name,atoi(pass),digits,window,&r,netkey));
	else r=neoauth(t,name,atoi(pass),slot1,slot2,window,digits,serial,
		neokey,tryusb,reset);
	if(r==PAM_SUCCESS&&replaydb)
		r=replaycheck(pamh,replaydb,rcmp,name,atoi(pass),t);
	if(r==PAM_SUCCESS&&cachedb)cacheadd(pamh,cachedb,cmp);

out:	config_clean();
	if(alwaysok)return PAM_SUCCESS;
	return r;
}

#ifdef PAM_STATIC

struct pam_module _pam_neototp_modstruct=
{
	"pam_neototp",
	pam_sm_authenticate,
	pam_sm_setcred,
	pam_sm_acct_mgmt,
	pam_sm_open_session,
	pam_sm_close_session,
	pam_sm_chauthtok
};

#endif
